### PR TITLE
Improve Node Field Query API

### PR DESCRIPTION
**Pull request by CoolSpy3**
**Description**
Allows accessing the internal fields of a proto hierarchy from a supervisor controller. There's still a lot of work to be done to integrate this into Webots, but now that I have a working proof-of-concept, I'm opening this PR to keep track of the remaining tasks.

This PR introduces the following new supervisor API functions:
* `wb_supervisor_node_get_proto`
* `wb_supervisor_proto_get_type_name`
* `wb_supervisor_proto_get_parent`
* `wb_supervisor_proto_get_field`
* `wb_supervisor_proto_get_field_by_index`
* `wb_supervisor_proto_get_number_of_fields`
* `wb_supervisor_proto_is_derived`

These act on the new type `WbProtoRef`.

Additionally, the following function has been added for `WbFieldRef` objects:
*  `wb_supervisor_field_get_actual_field`

Finally, the following functions have been renamed for clarity:
* `wb_supervisor_node_get_proto_field` -> `wb_supervisor_node_get_base_node_field`
* `wb_supervisor_node_get_proto_field_by_index` -> `wb_supervisor_node_get_base_node_field_by_index`
* `wb_supervisor_node_get_number_of_proto_fields` -> `wb_supervisor_node_get_number_of_base_node_fields`

**UPDATE (Aug 19): After looking through the usages of the above methods, I've noticed that the `wb_supervisor_node_get_field` (and related) methods are used much more than the `proto` variants. Thus, I've decided to keep its functionality the same in order to minimize the amount of code that will need to be updated in existing controllers. I've renamed the other methods appropriately (and updated the above text). To see the originals, check the edit history of this comment.**

**Related Issues**
This pull-request fixes issue #6354.

**Tasks**
  - [x] Allow Webots to keep track of internal proto parameters
  - [x] Allow internal proto parameters to be accessed from the C supervisor API
  - [x] Ensure that updates to internal proto fields are propagated to the supervisor
  - [x] Add tests for the new API methods
  - [x] Propagate supervisor API changes to other languages
    - [x] Update the C++ API
    - [x] Update the Java API
    - [x] Update the Python API
    - [x] Update the Matlab API
    - [x] Update `WbLanguage.cpp` with the new api methods
  - [x] Update references to the renamed API methods throughout the codebase
  - ~~[ ] Add a sample usage demo~~ **After looking at the existing samples, I found that there is currently no sample for and of the `wb_supervisor_node_get_proto_field` methods. Thus, I'm not sure what I good sample for a purely introspective API would look like. For the moment, I'll assume that the advice to write samples is indented for more simulation-facing features (e.g. devices).**
  - [x] Update ROS
    - [x] Update the ROS Controller (cyberbotics/webots_ros#151)
      - [x] Add services for the new API methods & update existing services
      - [x] Add tests for the new API methods
    - ~~[ ] Update the ROS2 Controller~~ **([Looks like this is automatically done by an action](https://github.com/cyberbotics/webots_ros2/blob/master/.github/workflows/libcontroller_updater.yml))**
      - ~~[ ] Update the webots submodule~~
  - [x] Update the documentation
    - [x] Update documentation relating to the renamed methods and clarify the distinction between them
    - [x] Add documentation for the new API methods
  - [x] Update the [changelog](https://github.com/cyberbotics/webots/blob/master/docs/reference/changelog-r2024.md)

**Documentation**
[Supervisor API](https://cyberbotics.com/doc/reference/supervisor?version=DeepBlueRobotics:feature-improve-node-field-query-api)

### CODE REVIEW

**Comment by CoolSpy3** [comment]
Alright, I think this is ready to be reviewed! Keep in mind that [the associated webots_ros PR](https://github.com/cyberbotics/webots_ros/pull/151) should probably be merged first, [so that the submodule can be updated to point to a commit on `master`](https://github.com/cyberbotics/webots/actions/runs/10605730567/job/29395115950?pr=6613#step:5:4608).

**Comment by omichel** [comment]
Sorry, I don't have time right now to review this PR and that's a big one. I would appreciate if someone else could review it. @brettle: would have time for this?

**Comment by CoolSpy3** [comment]
Note: The tests are currently failing due to #6654.

Also, for whomever ends up reviewing this, do you think [WbNode.cpp:1795](https://github.com/cyberbotics/webots/blob/bf1e0272be333428ed099823921bce8bacbe7d78/src/webots/vrml/WbNode.cpp#L1795) could delete field references being held by `WbNodeProtoInfo`? It seems like it might, but I haven't yet found a test case that causes that behavior to occur (that includes running the current tests [which *should* cover all scenarios] with hidden fields [instead of normal/unconnected fields]). I would think using a hidden field in `SolidProtoHierarchyInternal` would cause this, but it doesn't. (Conversely, the tests fail without the current change to [WbNode:1782](https://github.com/cyberbotics/webots/blob/bf1e0272be333428ed099823921bce8bacbe7d78/src/webots/vrml/WbNode.cpp#L1782)). If you don't want to investigate, I'm happy enough with saying "the tests should cover all use cases" and seeing if a bug pops up later.

**Comment by brettle** [comment]
> Sorry, I don't have time right now to review this PR and that's a big one. I would appreciate if someone else could review it. @brettle: would you have time for this?

I'll see what I can do but it might take a while.

File: src/controller/c/supervisor.c [hunk 1]
```diff
@@ -979,12 +1091,14 @@ static void supervisor_read_answer(WbDevice *d, WbRequest *r) {
...
+        const int field_proto_id = request_read_int32(r);
         const int field_id = request_read_int32(r);
         const bool is_field_get_request = sent_field_get_request && sent_field_get_request->field &&
                                           sent_field_get_request->field->node_unique_id == field_node_id &&
+                                          sent_field_get_request->field->proto_id == field_proto_id &&
                                           sent_field_get_request->field->id == field_id;
-        WbFieldStruct *f =
-          (is_field_get_request) ? sent_field_get_request->field : find_field_by_id(field_node_id, field_id, false);
+        WbFieldStruct *f = (is_field_get_request) ? sent_field_get_request->field :
+                                                    find_field_by_id(field_node_id, -1, field_id, false);  // TODO: FIX
```
Comment (brettle) [inline src/controller/c/supervisor.c:1101]: Is there still something to fix here or is the TODO no longer valid?

File: src/controller/c/supervisor.c [hunk 2]
```diff
@@ -949,12 +1037,32 @@ static void supervisor_read_answer(WbDevice *d, WbRequest *r) {
...
       }
     } break;
+    case C_SUPERVISOR_NODE_GET_PROTO: {
+      const int id = request_read_int32(r);
+      const bool is_derived = request_read_uchar(r) == 1;
+      const int num_fields = request_read_int32(r);
+      const char *type_name = request_read_string(r);
+      if (id < 0)
+        break;
+      WbProtoInfoStruct *p = malloc(sizeof(WbProtoInfoStruct));
```
Comment (brettle) [inline src/controller/c/supervisor.c:1047]: Isn't this a memory leak? Specifically, doesn't this result in a new `WbProtoInfoStruct` being allocated every time `wb_supervisor_node_get_proto` is called?

File: src/controller/c/supervisor.c (hunk 1 above)
Comment (CoolSpy3) [reply src/controller/c/supervisor.c]: Ahh, good catch! -1 should be replaced with the actual proto id. Fixed.

File: src/controller/c/supervisor.c (hunk 2 above)
Comment (CoolSpy3) [reply src/controller/c/supervisor.c:1047]: No, if `wb_supervisor_node_get_proto` has already been called, [the proto information will be cached in `node-proto_info`](https://github.com/cyberbotics/webots/blob/56a6fcf68bd6a94ba895000f9f6f06b51b92f935/src/controller/c/supervisor.c#L2928), and no request will be sent to Webots. The only exception is if the node is regenerated, in which case [the old proto structs are freed](https://github.com/cyberbotics/webots/blob/56a6fcf68bd6a94ba895000f9f6f06b51b92f935/src/controller/c/supervisor.c#L417https://github.com/cyberbotics/webots/blob/56a6fcf68bd6a94ba895000f9f6f06b51b92f935/src/controller/c/supervisor.c#L417).

Comment (brettle) [reply src/controller/c/supervisor.c:1047]: Nevermind. I see now that this is handled in `wb_supervisor_node_get_proto`.

File: src/controller/c/supervisor.c [hunk 3]
```diff
@@ -3675,3 +3918,157 @@ const char *wb_supervisor_field_get_type_name(WbFieldRef field) {
       return "";
   }
 }
+
+const char *wb_supervisor_proto_get_type_name(WbProtoRef proto) {
```
Comment (brettle) [inline src/controller/c/supervisor.c:3922]: The following functions don't seem to call `robot_check_supervisor`: `wb_supervisor_proto_get_type_name`, `wb_supervisor_proto_is_derived`, `wb_supervisor_proto_get_number_of_fields`. Is that intentional?

Comment (CoolSpy3) [reply src/controller/c/supervisor.c:3922]: For that, I was mirroring the `wb_supervisor_field_get_*` functions, which do not call `robot_check_supervisor` for values which can be obtained from the struct.

File: src/controller/cpp/Proto.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+#include <stdio.h>
+#include <map>
+
+using namespace std;
+using namespace webots;
+
+static map<WbProtoRef, Proto *> protoMap;
+
+Proto *Proto::findProto(WbProtoRef ref) {
```
Comment (brettle) [inline src/controller/cpp/Proto.cpp:27]: Consider changing name to `findOrAddProto`.

File: src/controller/cpp/Proto.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  if (!ref)
+    return NULL;
+
+  map<WbProtoRef, Proto *>::iterator iter = protoMap.find(ref);
+  if (iter != protoMap.end())
+    return iter->second;
+
+  Proto *field = new Proto(ref);
```
Comment (brettle) [inline src/controller/cpp/Proto.cpp:35]: Seems like `proto` would be a better name that `field` here.

File: src/controller/java/controller.i [hunk 6]
```diff
@@ -733,6 +771,57 @@ namespace webots {
...
+
+  public Field getFieldByIndex(int index) {
+    long cPtr = wrapperJNI.Proto_getFieldByIndexPrivate(swigCPtr, this, index);
+    return Field.findField(cPtr);
+  }
+
+  private static java.util.HashMap<Long,Proto> protos = new java.util.HashMap<Long,Proto>();
+
+  // DO NOT USE THIS FUNCTION: IT IS RESERVED FOR INTERNAL USE !
+  public static Proto findProto(long cPtr) {
```
Comment (brettle) [inline src/controller/java/controller.i:809]: Consider renaming to `findOrAddProto`.

File: src/controller/matlab/mgenerate.py [hunk 7]
```diff
@@ -517,28 +518,28 @@ def main(args=None):
...
+    generator.gen(FUNC, "wb_supervisor_node_get_base_node_field_by_index(noderef, fieldindex)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_base_type_name(noderef)", "supervisor")
     # generator.gen(FUNC, "wb_supervisor_node_get_center_of_mass(noderef)", "supervisor")
     # generator.gen(FUNC, "wb_supervisor_node_get_contact_point(noderef, index)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_contact_point_node(noderef, index)", "supervisor")
     # generator.gen(PROC, "wb_supervisor_node_get_contact_points(noderef, include_descendants)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_def(noderef)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_field(noderef, fieldname)", "supervisor")
-    generator.gen(FUNC, "wb_supervisor_node_get_field_by_index(noderef, fieldname)", "supervisor")
+    generator.gen(FUNC, "wb_supervisor_node_get_field_by_index(noderef, fieldindex)", "supervisor")
```
Comment (brettle) [inline src/controller/matlab/mgenerate.py:530]: Is this just a costmetic fix or does it change fix an actual bug and/or affect existing code? If it's not just costmetic, it should probably be added to the changelog.

File: src/controller/matlab/mgenerate.py [hunk 8]
```diff
@@ -517,28 +518,28 @@ def main(args=None):
...
     generator.gen(FUNC, "wb_supervisor_node_get_from_def(defname)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_from_device(tag)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_from_id(id)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_from_proto_def(noderef, defname)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_id(noderef)", "supervisor")
+    generator.gen(FUNC, "wb_supervisor_node_get_number_of_base_node_fields(noderef)", "supervisor")
     generator.gen(FUNC, "wb_supervisor_node_get_number_of_contact_points(noderef, include_descendants)", "supervisor")
-    generator.gen(FUNC, "wb_supervisor_node_get_number_of_fields(noderef, fieldname)", "supervisor")
+    generator.gen(FUNC, "wb_supervisor_node_get_number_of_fields(noderef)", "supervisor")
```
Comment (brettle) [inline src/controller/matlab/mgenerate.py:538]: Is this just a costmetic fix or does it change fix an actual bug and/or affect existing code? If it's not just costmetic, it should probably be added to the changelog.

File: src/controller/cpp/Proto.cpp (hunk 5 above)
Comment (CoolSpy3) [reply src/controller/cpp/Proto.cpp]: Yes it would! Fixed!

File: src/controller/cpp/Proto.cpp (hunk 4 above)
Comment (CoolSpy3) [reply src/controller/cpp/Proto.cpp:27]: The method name is modeled off of `Node::findNode` and `Field::fieldField`. However, I agree that that name is probably better. Do you think I should go through and change the others as well, just update this one, or leave it as is?

File: src/controller/java/controller.i (hunk 6 above)
Comment (CoolSpy3) [reply src/controller/java/controller.i:809]: [See comment in `Proto.cpp`](https://github.com/cyberbotics/webots/pull/6613#discussion_r1747772547).

File: src/controller/cpp/Proto.cpp (hunk 4 above)
Comment (brettle) [reply src/controller/cpp/Proto.cpp:27]: Leave it to stay consistent.

File: src/controller/matlab/mgenerate.py (hunk 7 above)
Comment (CoolSpy3) [reply src/controller/matlab/mgenerate.py:530]: Afaik, this is just cosmetic.

File: src/controller/matlab/mgenerate.py (hunk 8 above)
Comment (CoolSpy3) [reply src/controller/matlab/mgenerate.py:538]: Afaik, this is not purely cosmetic. I've added a changelog entry.

File: tests/api/controllers/supervisor_import_remove_mf/supervisor_import_remove_mf.c [hunk 9]
```diff
@@ -38,49 +38,55 @@ int main(int argc, char **argv) {
...
 
   wb_robot_step(TIME_STEP);
-
   wb_supervisor_field_insert_mf_bool(bool_field, 0, true);
   reference_distance -= 0.05;
   wb_robot_step(TIME_STEP);
   ts_assert_double_in_delta(wb_distance_sensor_get_value(distance_sensor), reference_distance, 0.0001,
                             "Distance after insertion of MFBool incorrect.");
 
+  wb_robot_step(TIME_STEP);
```
Comment (brettle) [inline tests/api/controllers/supervisor_import_remove_mf/supervisor_import_remove_mf.c:47]: Are the calls to wb_robot_step(TIME_STEP) that you've added to this file related to this PR in some way? If they are, please explain. If they aren't, please move them to another PR if they are needed to address some other issue.

File: tests/api/controllers/supervisor_notify_import_remove_mf/supervisor_notify_import_remove_mf.c [hunk 10]
```diff
@@ -12,38 +12,71 @@ int main(int argc, char **argv) {
...
+    mf_field[i] = wb_supervisor_node_get_field(node, field_names[i]);
+    mf_proto_fields[i] = wb_supervisor_proto_get_field(proto, field_names[i]);
+  }
+
+  int mf_field_count[FIELD_COUNT];
+  for (i = 0; i < FIELD_COUNT; ++i) {
+    const int count = wb_supervisor_field_get_count(mf_field[i]);
+    mf_field_count[i] = count;
+    const int proto_count = wb_supervisor_field_get_count(mf_proto_fields[i]);
+    ts_assert_int_equal(proto_count, count, "Size of proto field %d not correctly initialized: found %d, expected %d", i, count,
```
Comment (brettle) [inline tests/api/controllers/supervisor_notify_import_remove_mf/supervisor_notify_import_remove_mf.c:30]: Presumably, the last 3 params should be "i, proto_count, count"?

File: tests/api/controllers/supervisor_import_remove_mf/supervisor_import_remove_mf.c (hunk 9 above)
Comment (brettle) [reply tests/api/controllers/supervisor_import_remove_mf/supervisor_import_remove_mf.c:47]: Nevermind. I understand now.

File: tests/api/controllers/supervisor_notify_import_remove_mf/supervisor_notify_import_remove_mf.c (hunk 10 above)
Comment (CoolSpy3) [reply tests/api/controllers/supervisor_notify_import_remove_mf/supervisor_notify_import_remove_mf.c]: That is correct; fixed!

**Comment by brettle** [comment]
> Also, for whomever ends up reviewing this, do you think [WbNode.cpp:1795](https://github.com/cyberbotics/webots/blob/bf1e0272be333428ed099823921bce8bacbe7d78/src/webots/vrml/WbNode.cpp#L1795) could delete field references being held by `WbNodeProtoInfo`? It seems like it might, but I haven't yet found a test case that causes that behavior to occur (that includes running the current tests [which _should_ cover all scenarios] with hidden fields [instead of normal/unconnected fields]). I would think using a hidden field in `SolidProtoHierarchyInternal` would cause this, but it doesn't. (Conversely, the tests fail without the current change to [WbNode:1782](https://github.com/cyberbotics/webots/blob/bf1e0272be333428ed099823921bce8bacbe7d78/src/webots/vrml/WbNode.cpp#L1782)). If you don't want to investigate, I'm happy enough with saying "the tests should cover all use cases" and seeing if a bug pops up later.

I'm ok with relying on the test coverage.

**Review by brettle — APPROVED** [review]
Excellent work!

File: docs/reference/supervisor.md [hunk 11]
```diff
@@ -4012,6 +4024,206 @@ rootChildrenField.importMFNodeFromString(4, 'DEF MY_ROBOT Robot { controller "my
...
+#### `wb_supervisor_proto_get_parent`
+
+%tab-component "language"
+
+%tab "C"
+
+```c
+#include <webots/supervisor.h>
+
+const char *wb_supervisor_proto_get_type_name(WbNodeRef proto);
```
Comment (omichel) [inline docs/reference/supervisor.md:4038]: I believe there is an error here and at several other places in the documentation:
```suggestion
const char *wb_supervisor_proto_get_type_name(WbProtoRef proto);
```

Comment (CoolSpy3) [reply docs/reference/supervisor.md]: Fixed! I also added the missing documentation for `wb_supervisor_node_get_proto`.

**Review by omichel — APPROVED** [review]
Thank you!
And thanks also to @brettle for the nice review!

**Pull request merged**